packages/gatsby/src/utils/webpack/static-query-mapper.ts

```diff
@@ -0,0 +1,260 @@
+              chunkGroupsDerivedFromEntrypoints.push(chunkGroup)
+            } else {
+              chunkGroupsDerivedFromEntrypoints = chunkGroupsDerivedFromEntrypoints.concat(
+                getChunkGroupsDerivedFromEntrypoint(chunkGroup, appEntryPoint)
+              )
+            }
+          }
+        }
+
+        // loop over all component chunkGroups or global ones
+        chunkGroupsDerivedFromEntrypoints.forEach(chunkGroup => {
+          const staticQueryHashes =
+            staticQueriesByChunkGroup.get(chunkGroup) ?? []
+
+          staticQueryHashes.push(
+            (staticQueryComponents.get(
+              staticQueryId
+            ) as IGatsbyStaticQueryComponents).hash
+          )
+
+          staticQueriesByChunkGroup.set(chunkGroup, staticQueryHashes)
+        })
+      }
+
+      // group chunkGroups by componentPaths for ease of use
+      for (const [
+        componentPath,
+        webpackModule,
+      ] of webpackModulesByComponentId) {
+        for (const chunk of compilation.chunkGraph.getModuleChunksIterable(
+          webpackModule
+        )) {
+          for (const chunkGroup of chunk.groupsIterable) {
+            // When it's a direct import from app entrypoint (async-requires) we know we have the correct chunkGroup
+            if (chunkGroup.name === generateComponentChunkName(componentPath)) {
+              chunkGroupsWithPageComponents.add(chunkGroup)
+              chunkGroupsByComponentPath.set(componentPath, chunkGroup)
+            }
+          }
+        }
+      }
+
+      let globalStaticQueries: Array<string> = []
+      for (const [chunkGroup, staticQueryHashes] of staticQueriesByChunkGroup) {
+        // When a chunkgroup is not part of a pageComponent we know it's part of a global group.
+        if (!chunkGroupsWithPageComponents.has(chunkGroup)) {
+          globalStaticQueries = globalStaticQueries.concat(staticQueryHashes)
+        }
+      }
+
+      components.forEach(component => {
+        const allStaticQueries = new Set(globalStaticQueries)
+        if (chunkGroupsByComponentPath.has(component.componentPath)) {
+          const chunkGroup = chunkGroupsByComponentPath.get(
+            component.componentPath
+          )
+          if (chunkGroup && staticQueriesByChunkGroup.has(chunkGroup)) {
+            ;(staticQueriesByChunkGroup.get(chunkGroup) as Array<
+              string
+            >).forEach(staticQuery => {
+              allStaticQueries.add(staticQuery)
+            })
+          }
+        }
+
+        this.store.dispatch({
+          type: `ADD_PENDING_TEMPLATE_DATA_WRITE`,
+          payload: {
+            componentPath: component.componentPath,
+            pages: component.pages,
+          },
+        })
+
+        this.store.dispatch({
+          type: `SET_STATIC_QUERIES_BY_TEMPLATE`,
+          payload: {
+            componentPath: component.componentPath,
+            staticQueryHashes: Array.from(allStaticQueries),
+          },
+        })
```

Comment:
1. This should not be dispatched if static query used by template didn't change - otherwise we will be rewriting stuff unnecesairly - previously we had:
```js
            if (
              !isEqual(
                state.staticQueriesByTemplate.get(componentPath),
                staticQueryHashes
              )
            ) {
            // dispatch
            }
```
we don't have it anymore

2. `Iterables` we use from `chunkGraph` don't ensure deterministic order which means that `allStaticQueries` order is not stable and can swap between builds - we should do `.sort()` on it (before we do comparison from step 1)